### PR TITLE
crimson/zns: refactor zone operations

### PR DESCRIPTION
**Pull request by aravind-wdc**
Moved all zone operations into one routine.

Signed-off-by: Aravind Ramesh <Aravind.Ramesh@wdc.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/crimson/os/seastore/segment_manager/zns.cc [hunk 1]
```diff
@@ -395,26 +395,57 @@ struct blk_zone_range make_range(
...
+		            blk_zone_range &range,
+			    zone_op_t zone_op){
+  LOG_PREFIX(ZNSSegmentManager::blk_zone_op);
+
+  unsigned long ioctl_op = 0;
+  std::string ioctl_name = "";
+  switch (zone_op) {
+    case zone_op_t::ZONE_OPEN:
+      ioctl_op = BLKOPENZONE;
+      ioctl_name = "BLKOPENZONE";
```
Comment (tchaikov) [inline src/crimson/os/seastore/segment_manager/zns.cc:411]: better off specialize a `fmt::formatter<zone_op_t>` . so we don't need to 1) create a `std::string` , 2) call its assignment operator and then 3) capture it by value, only for the failure case.

File: src/crimson/os/seastore/segment_manager/zns.cc [hunk 2]
```diff
@@ -395,26 +395,57 @@ struct blk_zone_range make_range(
...
 
-using blk_open_zone_ertr = crimson::errorator<
+using blk_zone_op_ertr = crimson::errorator<
   crimson::ct_error::input_output_error>;
-using blk_open_zone_ret = blk_open_zone_ertr::future<>;
-blk_open_zone_ret blk_open_zone(seastar::file &device, blk_zone_range &range){
+using blk_zone_op_ret = blk_zone_op_ertr::future<>;
+blk_zone_op_ret blk_zone_op(seastar::file &device,
+		            blk_zone_range &range,
+			    zone_op_t zone_op){
```
Comment (cyx1231st) [inline src/crimson/os/seastore/segment_manager/zns.cc:403]: `) {`

File: src/crimson/os/seastore/segment_manager/zns.cc [hunk 3]
```diff
@@ -395,26 +395,57 @@ struct blk_zone_range make_range(
...
-    BLKOPENZONE, 
+    ioctl_op,
     &range
-  ).then_wrapped([=](auto f) -> blk_open_zone_ret{
+  ).then_wrapped([=](auto f) -> blk_zone_op_ret {
     if (f.failed()) {
+      DEBUG("{} ioctl failed", ioctl_name.c_str());
       return crimson::ct_error::input_output_error::make();
     }
     else {
```
Comment (cyx1231st) [inline src/crimson/os/seastore/segment_manager/zns.cc:438]: `} else {`

File: src/crimson/os/seastore/segment_manager/zns.cc [hunk 4]
```diff
@@ -395,26 +395,57 @@ struct blk_zone_range make_range(
...
       int ret = f.get();
       if (ret == 0) {
 	return seastar::now();
       } else {
+        DEBUG("{} ioctl failed with return code {}", ioctl_name.c_str(), ret);
 	return crimson::ct_error::input_output_error::make();
       }
     }
   });
+
```
Comment (cyx1231st) [inline src/crimson/os/seastore/segment_manager/zns.cc:448]: unrelated change.

File: src/crimson/os/seastore/segment_manager/zns.cc [hunk 5]
```diff
@@ -395,26 +395,57 @@ struct blk_zone_range make_range(
...
+  }
+
   return device.ioctl(
-    BLKOPENZONE, 
+    ioctl_op,
     &range
-  ).then_wrapped([=](auto f) -> blk_open_zone_ret{
+  ).then_wrapped([=](auto f) -> blk_zone_op_ret {
     if (f.failed()) {
+      DEBUG("{} ioctl failed", ioctl_name.c_str());
```
Comment (cyx1231st) [inline src/crimson/os/seastore/segment_manager/zns.cc:435]: better to be error log?

File: src/crimson/os/seastore/segment_manager/zns.cc [hunk 6]
```diff
@@ -395,26 +395,57 @@ struct blk_zone_range make_range(
...
       return crimson::ct_error::input_output_error::make();
     }
     else {
       int ret = f.get();
       if (ret == 0) {
 	return seastar::now();
       } else {
+        DEBUG("{} ioctl failed with return code {}", ioctl_name.c_str(), ret);
```
Comment (cyx1231st) [inline src/crimson/os/seastore/segment_manager/zns.cc:443]: better to be error log?

File: src/crimson/os/seastore/segment_manager/zns.cc (hunk 1 above)
Comment (aravind-wdc) [reply src/crimson/os/seastore/segment_manager/zns.cc]: Thanks for the review.
Added fmt::formatter<zone_op_t>.

File: src/crimson/os/seastore/segment_manager/zns.cc (hunk 2 above)
Comment (aravind-wdc) [reply src/crimson/os/seastore/segment_manager/zns.cc]: Done

File: src/crimson/os/seastore/segment_manager/zns.cc (hunk 3 above)
Comment (aravind-wdc) [reply src/crimson/os/seastore/segment_manager/zns.cc]: Done.

File: src/crimson/os/seastore/segment_manager/zns.cc (hunk 4 above)
Comment (aravind-wdc) [reply src/crimson/os/seastore/segment_manager/zns.cc]: Done.

File: src/crimson/os/seastore/segment_manager/zns.cc (hunk 5 above)
Comment (aravind-wdc) [reply src/crimson/os/seastore/segment_manager/zns.cc]: Yes, changed.

File: src/crimson/os/seastore/segment_manager/zns.cc (hunk 6 above)
Comment (aravind-wdc) [reply src/crimson/os/seastore/segment_manager/zns.cc]: Thanks for the review.

File: src/crimson/os/seastore/segment_manager/zns.cc [hunk 7]
```diff
@@ -17,6 +18,29 @@ SET_SUBSYS(seastore_device);
 // limit the max padding buf size to 1MB
 #define MAX_PADDING_SIZE 1048576
 
+using z_op = crimson::os::seastore::segment_manager::zns::zone_op_t;
```
Comment (tchaikov) [inline src/crimson/os/seastore/segment_manager/zns.cc:21]: just in case.. now that we are using C++20. we could use something like:

```c++
template <>
struct fmt::formatter<crimson::os::seastore::segment_manager::zns::zone_op_t> : fmt::formatter<std::string_view> {
  using enum crimson::os::seastore::segment_manager::zns::zone_op_t;
  std::string_view name = "Unknown";
  switch (s) {
  case ZONE_OPEN:
    name = "BLKOPENZONE";
    break;
  // ...
  }
};
```

File: src/crimson/os/seastore/segment_manager/zns.h [hunk 8]
```diff
@@ -63,6 +63,13 @@ namespace crimson::os::seastore::segment_manager::zns {
   using write_ertr = crimson::errorator<crimson::ct_error::input_output_error>;
   using read_ertr = crimson::errorator<crimson::ct_error::input_output_error>;
 
+  enum class zone_op_t {
+    ZONE_OPEN,
+    ZONE_FINISH,
+    ZONE_CLOSE,
+    ZONE_RESET,
```
Comment (tchaikov) [inline src/crimson/os/seastore/segment_manager/zns.h:70]: might want to drop the "ZONE" prefix.

File: src/crimson/os/seastore/segment_manager/zns.h [hunk 9]
```diff
@@ -63,6 +63,13 @@ namespace crimson::os::seastore::segment_manager::zns {
   using write_ertr = crimson::errorator<crimson::ct_error::input_output_error>;
   using read_ertr = crimson::errorator<crimson::ct_error::input_output_error>;
 
+  enum class zone_op_t {
```
Comment (tchaikov) [inline src/crimson/os/seastore/segment_manager/zns.h:66]: nit, s/zone_op_t/zone_op/.

File: src/crimson/os/seastore/segment_manager/zns.cc [hunk 10]
```diff
@@ -395,22 +419,46 @@ struct blk_zone_range make_range(
...
 
-using blk_open_zone_ertr = crimson::errorator<
+using blk_zone_op_ertr = crimson::errorator<
   crimson::ct_error::input_output_error>;
-using blk_open_zone_ret = blk_open_zone_ertr::future<>;
-blk_open_zone_ret blk_open_zone(seastar::file &device, blk_zone_range &range){
+using blk_zone_op_ret = blk_zone_op_ertr::future<>;
+blk_zone_op_ret blk_zone_op(seastar::file &device,
+		            blk_zone_range &range,
+			    zone_op_t zone_op) {
```
Comment (tchaikov) [inline src/crimson/os/seastore/segment_manager/zns.cc:427]: s/zone_op/op/

**Review by tchaikov — COMMENTED** [review]
just some nits.

File: src/crimson/os/seastore/segment_manager/zns.cc (hunk 10 above)
Comment (aravind-wdc) [reply src/crimson/os/seastore/segment_manager/zns.cc:427]: Ok, will change it.

File: src/crimson/os/seastore/segment_manager/zns.h (hunk 9 above)
Comment (aravind-wdc) [reply src/crimson/os/seastore/segment_manager/zns.h:66]: Ok, I saw that in most of the places enums was defined like this(abc_t), so I followed the existing conventions.
I will change it.

File: src/crimson/os/seastore/segment_manager/zns.h (hunk 8 above)
Comment (aravind-wdc) [reply src/crimson/os/seastore/segment_manager/zns.h:70]: Ok

File: src/crimson/os/seastore/segment_manager/zns.cc (hunk 7 above)
Comment (aravind-wdc) [reply src/crimson/os/seastore/segment_manager/zns.cc:21]: Initially I did like this, but then the lines felt un-necessarily long, so I used `using z_op = ...`
Will change it.

Comment (tchaikov) [reply src/crimson/os/seastore/segment_manager/zns.cc:21]: no problem. just wanted to show that we could use `using enum` to expose the enum members. if you prefer a shorter name, probably we could use both of them, namely the type alias and `using enum`.
```c++
using zone_op = crimson::os::seastore::segment_manager::zns::zone_op_t;
/// ...
using enum zone_op;
```

File: src/crimson/os/seastore/segment_manager/zns.h (hunk 9 above)
Comment (tchaikov) [reply src/crimson/os/seastore/segment_manager/zns.h:66]: yeah, i believe the `_t` postfix dated back to the old days when we still need to `typedef` an enum for using it with the typedef'ed name. but since we are using C++, there is no need to typedef an enum anymore.

File: src/crimson/os/seastore/segment_manager/zns.cc (hunk 7 above)
Comment (aravind-wdc) [reply src/crimson/os/seastore/segment_manager/zns.cc:21]: Yes, this feels better. Updated. Thanks

File: src/crimson/os/seastore/segment_manager/zns.h (hunk 9 above)
Comment (aravind-wdc) [reply src/crimson/os/seastore/segment_manager/zns.h:66]: Ok, updated with the new changes.

File: src/crimson/os/seastore/segment_manager/zns.h (hunk 8 above)
Comment (aravind-wdc) [reply src/crimson/os/seastore/segment_manager/zns.h:70]: Updated.

File: src/crimson/os/seastore/segment_manager/zns.cc (hunk 10 above)
Comment (aravind-wdc) [reply src/crimson/os/seastore/segment_manager/zns.cc:427]: Updated.

**Comment by aravind-wdc** [comment]
Rebased.

File: src/crimson/os/seastore/segment_manager/zns.cc [hunk 11]
```diff
@@ -397,22 +422,47 @@ struct blk_zone_range make_range(
...
+      ioctl_op = BLKFINISHZONE;
+      break;
+    case RESET:
+      ioctl_op = BLKRESETZONE;
+      break;
+    case CLOSE:
+      ioctl_op = BLKCLOSEZONE;
+      break;
+    default:
+      ERROR("Invalid zone operation {}", fmt::format("{}", op));
```
Comment (cyx1231st) [inline src/crimson/os/seastore/segment_manager/zns.cc:449]: `ERROR("Invalid zone operation {}", op);`

File: src/crimson/os/seastore/segment_manager/zns.cc [hunk 12]
```diff
@@ -397,22 +422,47 @@ struct blk_zone_range make_range(
...
+  }
+
   return device.ioctl(
-    BLKOPENZONE, 
+    ioctl_op,
     &range
-  ).then_wrapped([=](auto f) -> blk_open_zone_ret{
+  ).then_wrapped([=](auto f) -> blk_zone_op_ret {
     if (f.failed()) {
+      ERROR("{} ioctl failed", fmt::format("{}", op));
```
Comment (cyx1231st) [inline src/crimson/os/seastore/segment_manager/zns.cc:458]: `ERROR("{} ioctl failed", op);`

File: src/crimson/os/seastore/segment_manager/zns.cc [hunk 13]
```diff
@@ -397,22 +422,47 @@ struct blk_zone_range make_range(
...
       return crimson::ct_error::input_output_error::make();
-    }
-    else {
+    } else {
       int ret = f.get();
       if (ret == 0) {
 	return seastar::now();
       } else {
+        ERROR("{} ioctl failed with return code {}", fmt::format("{}", op), ret);
```
Comment (cyx1231st) [inline src/crimson/os/seastore/segment_manager/zns.cc:465]: `ERROR("{} ioctl failed with return code {}", op, ret);`

File: src/crimson/os/seastore/segment_manager/zns.cc (hunk 11 above)
Comment (aravind-wdc) [reply src/crimson/os/seastore/segment_manager/zns.cc:449]: Fixed.

File: src/crimson/os/seastore/segment_manager/zns.cc (hunk 12 above)
Comment (aravind-wdc) [reply src/crimson/os/seastore/segment_manager/zns.cc:458]: Fixed.

File: src/crimson/os/seastore/segment_manager/zns.cc (hunk 13 above)
Comment (aravind-wdc) [reply src/crimson/os/seastore/segment_manager/zns.cc:465]: Thanks. Fixed.

**Review by cyx1231st — APPROVED** [review]

**Review by tchaikov — APPROVED** [review]
